proposals/0072-feature-gate-threshold-automation.md

```diff
@@ -0,0 +1,193 @@
+
+## Detailed Design
+
+The new process is comprised of the following steps:
+
+1. **Feature Creation:** Contributors create feature accounts as they do now.
+2. **Staging Features for Activation:** In some epoch N, a multi-signature
+   authority stages for activation some or all of the features created in step
+   1, to be activated at the end of the *next epoch*.
+3. **Signaling Support for Staged Features:** During the next epoch (epoch N+1),
+   validators signal which of the staged feature-gates they support in their
+   software.
+4. **Feature Activation:** At the end of epoch N+1, the runtime activates the
+   feature-gates that have the required stake support.
+
+### Step 1: Feature Creation
+
+The first step is creation of a feature account, done by submitting a
+transaction containing System instructions to fund, allocate, and assign the
+feature account to `Feature111111111111111111111111111111111111`.
+
+This step is unchanged from its original procedure.
+
+### Step 2: Staging Features for Activation
+
+A multi-signature authority, comprised of key-holders from Solana Labs and
+possibly other validator client teams in the future, will have the authority to
+stage created features for activation.
+In the future, this authority could be replaced by validator governance.
+
+The multi-signature authority stages a feature for activation by invoking the
+Feature Gate program's `StageFeatureForActivation` instruction. This instruction
+expects:
+
+- Data: The feature ID
+- Accounts:
+  - Staged Features PDA: writable
+  - Multi-signature authority: signer 
+
+One single PDA will be used to store two lists of features. Its data will be
+structured as follows:
+
+- 8 bytes for the current epoch.
+- A fixed-size list of size 8 containing 32-byte feature IDs staged for
+  activation at the end of the *current epoch*.
+- A fixed-size list of size 8 containing the `u64` value of total stake support
+  for each feature ID in the prior list.
+- 8 bytes for the next epoch.
+- A fixed-size list of size 8 containing 32-byte feature IDs staged for
+  activation at the end of the *next epoch*.
+
+`StageFeatureForActivation` will add the provided feature ID to the **next
+epoch's** set of staged features.
+
+An example of how this account's state might work to coordinate both lists can
+be found here:
+
+<https://github.com/buffalojoec/feature-gate-data-poc/blob/master/src/lib.rs>
+
+The Staged Features PDA could be derived simply from one literal seed:
+`"staged_features"`.
+
+### Step 3: Signaling Support for Staged Features
+
+With an on-chain reference point to determine the features staged for activation
+for a particular epoch, nodes can signal their support for the staged features
+supported by their software.
+
+A node signals its support for staged features by invoking the Feature Gate
+program's `SignalSupportForStagedFeatures` instruction. This instruction
+expects:
+
+- Data: A `u8` bit mask of the staged features.
```

Comment:
For completeness, do you mind explaining the reasoning for using a bit mask here? Is it to save on block space/network bandwidth?

Reply:
Bytes over the wire and program compute, really. Removes the need to send a list of public keys and then task the program with searching the list. The bitmask is already in order.

I can add this explanation.

Reply:
Awesome thank-you!